src/common/src/util/epoch.rs

```diff
@@ -70,6 +70,16 @@ impl Epoch {
     pub fn as_system_time(&self) -> SystemTime {
         *UNIX_SINGULARITY_DATE_EPOCH + Duration::from_millis(self.physical_time())
     }
+
+    pub fn min_epoch_with_interval(&self, interval: u64) -> Self {
```

Comment:
Calling it `min_epoch_with_interval` is a bit confusing since the Epoch struct should be unaware of the usage of epoch. I suggest rename it to something like `subtract_ms(&self, relative_time_ms: u64)` and add some docs explaining the usage.